core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushFilterIntoValues.java

```diff
@@ -0,0 +1,165 @@
+    {
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isPushFilterIntoValues(session);
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        ValuesNode valuesNode = captures.get(VALUES);
+
+        Expression predicate = node.getPredicate();
+        if (!isDeterministic(predicate)) {
+            return Result.empty();
+        }
+
+        if (valuesNode.getRows().orElseThrow().isEmpty()) {
+            // empty values node, filter is redundant
+            return Result.ofPlanNode(valuesNode);
+        }
+
+        if (valuesNode.getRows().orElseThrow().stream()
+                .anyMatch(row -> !isDeterministic(row))) {
+            return Result.empty();
+        }
+
+        Set<Symbol> valuesExpressionsSymbols = valuesNode.getRows().orElseThrow().stream()
+                .map(SymbolsExtractor::extractUnique)
+                .flatMap(Set::stream)
+                .collect(toImmutableSet());
+        if (!valuesExpressionsSymbols.isEmpty()) {
+            // unresolved correlation
+            return Result.empty();
+        }
+
+        Set<Symbol> valuesOutputSymbols = ImmutableSet.copyOf(valuesNode.getOutputSymbols());
+        Set<Symbol> predicateSymbols = extractUnique(predicate);
+        if (!valuesOutputSymbols.containsAll(predicateSymbols)) {
+            // filter is correlated
+            return Result.empty();
+        }
+
+        ImmutableList.Builder<Expression> filteredRows = ImmutableList.builder();
+        boolean optimized = false;
+        boolean keepFilter = false;
+        for (Expression expression : valuesNode.getRows().orElseThrow()) {
+            Row row = (Row) expression;
+            ImmutableMap.Builder<Symbol, Expression> mapping = ImmutableMap.builder();
+            for (int i = 0; i < valuesNode.getOutputSymbols().size(); i++) {
+                mapping.put(valuesNode.getOutputSymbols().get(i), row.items().get(i));
+            }
+            Expression rewrittenPredicate = inlineSymbols(mapping.buildOrThrow(), predicate);
+            Expression optimizedPredicate = new IrExpressionInterpreter(rewrittenPredicate, plannerContext, context.getSession()).optimize();
+            if (optimizedPredicate.equals(TRUE)) {
+                filteredRows.add(row);
+            }
+            else if (optimizedPredicate.equals(FALSE) || optimizedPredicate.equals(NULL_BOOLEAN)) {
+                // skip row
+                optimized = true;
+            }
+            else {
+                // could not evaluate the predicate for the row
+                filteredRows.add(row);
+                keepFilter = true;
+            }
+        }
+
+        if (!optimized && keepFilter) {
+            return Result.empty();
```

Comment:
if the rule application can be expensive, would it be interesting to somehow prevent it from being applied again over same input (eg if some rule fires somewhere else)?
Perhaps something inspired by https://github.com/trinodb/trino/blob/35bc11ec9e37eb489c996d10f71377fd0c372acd/core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java#L128

Reply:
Let me address the concerns about cost with a modified session property. Instead of the on/off switch, let's define the max size of VALUES for which the rule fires, and set it to 100 rows by default.

Reply:
for bigger VALUES it still may make sense, but perhaps we may need a compiled filter

Reply:
I believe users shouldn't use much bigger values (like 1000 rows). Even if we adjust this rule to evaluate the filter more efficiently, the values will still be optimized by the interpreter as part of other optimizations (`SimplifyExpressions`), and eventually they will be evaluated in the `LocalExecutionPlanner`. I have seen queries timing out in the planner after 10 minutes because of too big values. 

Can we have this rule in its current shape?

Reply:
Good point